src/core/channelz/channelz.h

```diff
@@ -104,12 +106,39 @@ class BaseNode : public RefCounted<BaseNode> {
   intptr_t uuid() const { return uuid_; }
   const std::string& name() const { return name_; }
 
+ protected:
+  void PopulateJsonFromDataSources(Json::Object& json);
+
  private:
   // to allow the ChannelzRegistry to set uuid_ under its lock.
   friend class ChannelzRegistry;
+  // allow data source to register/unregister itself
+  friend class DataSource;
   const EntityType type_;
   intptr_t uuid_;
   std::string name_;
+  Mutex data_sources_mu_;
+  absl::InlinedVector<DataSource*, 3> data_sources_
+      ABSL_GUARDED_BY(data_sources_mu_);
+};
+
+class DataSource {
+ public:
+  explicit DataSource(RefCountedPtr<BaseNode> node);
+
+  // Add any relevant json fragments to the output.
+  // This method must not cause the DataSource to be deleted, or else there will
+  // be a deadlock.
+  virtual void AddJson(Json::Object& output) = 0;
+
+ protected:
+  ~DataSource();
```

Comment:
I think this needs to be virtual.

Reply:
it doesn't because we never delete a data source from the base pointer - so a protected non-virtual destructor is appropriate here

Reply:
Okay.  Please add a comment documenting that.